File: leetcodes/238. Product of Array Except Self/productExceptSelf.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
 var productExceptSelf = function(nums) {
    //res array -> left shift * right shift = final res
    //time O(2n) , space O(1) (don't count output array)
    let ans = [];
    let right = 1;
    let left = 1;
    for (let i = 0; i < nums.length; i++) {
      if (!ans[i]) ans[i] = 1;
      ans[i] *= right;
      right *= nums[i];
    }
    for (let j = nums.length - 1; j >= 0; j--) {
      ans[j] *= left;
      left *= nums[j];
    }
    return ans;
};

File: leetcodes/238. Product of Array Except Self/productExceptSelf.java
class Solution {
  public int[] productExceptSelf(int[] nums) {
      int lenN = nums.length;
      int[] res = new int[lenN];
      int temp = 1;
      for (int i = 0; i < lenN; i++) {
        if (res[i]== 0) res[i] = 1;
        res[i] *= temp;
        temp *= nums[i];
      }
      temp = nums[lenN - 1];
      for (int j = lenN - 1; j >= 0; j --) {
        res[j] *= temp;
        temp *= nums[j];
      }
      return res;
  }
}
//time O(2n) spaceO(1): right anad left two ways